src/timearray.jl

```diff
@@ -69,13 +69,67 @@ length(ata::AbstractTimeSeries) = length(ata.timestamp)
 
 size(ta::TimeArray, dim...) = size(ta.values, dim...)
 
-###### iterator protocol #########
+###### iterator protocol ########
 
 start(ta::TimeArray)   = 1
 next(ta::TimeArray, i) = ((ta.timestamp[i], ta.values[i, :]), i + 1)
 done(ta::TimeArray, i) = (i > length(ta))
 isempty(ta::TimeArray) = (length(ta) == 0)
 
+###### equal ####################
+
+"""
+    ==(x::TimeArray, y::TimeArray)
+
+If `true`, all fields of `x` and `y` should be equal,
+meaning that the two `TimeArray`s have the same values at the same points in time,
+the same colnames and the same metadata.
+
+Implies
+
+```julia
+x.timestamp == y.timestamp &&
+x.values    == y.values    &&
+x.colnames  == y.colnames  &&
+x.meta      == y.meta
+```
+"""
+@generated function ==(x::TimeArray{T,N}, y::TimeArray{S,M}) where {T,S,N,M}
+    if N != M
+        return :false
+        # Other type info is not helpful for assertion.
+        # e.g.
+        #      1.0 == 1
+        #      Date(2111, 1, 1) == DateTime(2111, 1, 1)
+    end
+
+    quote
+        for f ∈ fieldnames(TimeArray)
+            getfield(x, f) != getfield(y, f) && return :false
+        end
+        :true
+    end
+end
+
+@generated function isequal(x::TimeArray{T,N}, y::TimeArray{S,M}) where {T,S,N,M}
+    if N != M
+      return :false
+    end
+
+    quote
+        for f ∈ fieldnames(TimeArray)
+          !isequal(getfield(x, f), getfield(y, f)) && return :false
+        end
+        :true
+    end
+end
+
+# support for Dict
+hash(x::TimeArray, h::UInt) =
+    mapreduce(+, fieldnames(TimeArray)) do f
```

Comment:
`sum` is `mapreduce(+, ...)` so this can be written
```julia
Base.hash(x::TimeArray, h::UInt) = sum(f->hash(getfield(x, f), h), fieldnames(TimeArray))
```

Reply:
done